# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1511125265 18000
#      Sun Nov 19 16:01:05 2017 -0500
# Node ID 81557c74be7f5347e2cb8792af50ee8b2cbbfea4
# Parent  a01c210e42874301ee870478c337142f5beb11a7
improve error message for invalid anonymous functions

* oct-parse.in.yy (anon_fcn_handle): Improve error message produced
for anonymous function parse errors.

diff --git a/libinterp/parse-tree/oct-parse.in.yy b/libinterp/parse-tree/oct-parse.in.yy
--- a/libinterp/parse-tree/oct-parse.in.yy
+++ b/libinterp/parse-tree/oct-parse.in.yy
@@ -603,16 +603,24 @@ fcn_handle      : '@' FCN_HANDLE
                   }
                 ;
 
 anon_fcn_handle : '@' param_list stmt_begin expr_no_assign
                   {
                     $$ = parser.make_anon_fcn_handle ($2, $4);
                     lexer.nesting_level.remove ();
                   }
+                | '@' param_list stmt_begin error
+                  {
+                    YYUSE ($2);
+
+                    $$ = nullptr;
+                    parser.bison_error ("anonymous function bodies must be single expressions");
+                    YYABORT;
+                  }
                 ;
 
 primary_expr    : identifier
                   { $$ = $1; }
                 | constant
                   { $$ = $1; }
                 | fcn_handle
                   { $$ = $1; }
